### ISSUE TITLE
Keybindings should be configurable

### ISSUE THREAD

**Issue by daleglass** [vircadia/vircadia#807]
[New user on Gaming on Linux ](https://www.gamingonlinux.com/2020/10/vircadia-is-an-in-development-free-and-open-source-decentralized-3d-social-space/page=1#r191225) complained that key bindings aren't configurable and fixed on QWERTY.

**Comment by HifiExperiments** [comment]
I think this might actually be possible with the JS API, to some degree?  the initial keybindings (at least, for a lot of movement) is defined in a JSON file, and I think there's a JS API for setting up controller mappings that could override that.  someone would have to write an app to actual handle the mapping and we'd have to have some way to save it across runs though

**Comment by kasenvr** [comment]
Would it not be possible to stringify the JSON and save it to the interface.json?

**Comment by HifiExperiments** [comment]
yeah but the mapping that would be done in the script wouldn't be JSON internally, you'd need to convert it to JSON in a way that the script knows how to parse (all possible)

**Comment by SilverfishVR** [comment]
yes you can remap just about anything with the controller api, but weirdly the method to load from a json file [loadMapping](https://apidocs.vircadia.dev/Controller.html#.loadMapping) is not implemented (or the docs are wrong)

not exactly user friendly, but, then again, I expect a linux user with a non-qwerty keyboard to be above average tech-savvy 😄

**Comment by kasenvr** [comment]
Well the idea is one of us will make an app to interface that given all APIs exist or all APIs that need to be made will be added to make that possible.

**Comment by ctrlaltdavid** [comment]
Confirmed in the C++: `Controller.loadMapping()` isn't implemented yet.

**Comment by kasenvr** [comment]
That might be a necessary step... :D

**Comment by daleglass** [comment]
still a thing.

**Comment by Penguin-Guru** [comment]
This is on my feature wishlist. If it has not been done by the time I get a new computer, I will probably attempt it. Can't say when that will be and can't promise I will succeed in my attempt. There are plenty of things on my list though so no need to wait for me. :)

**Comment by Penguin-Guru** [comment]
I've been fiddling with this over the past couple days. The commit history is a complete mess because I've been learning QML as I go but feel free to take a look if anyone's interested. Current issue is that "Value is undefined and could not be converted to an object".

https://github.com/Penguin-Guru/vircadia/tree/InputConfiguration
